### ISSUE TITLE
Strategies for improving performance

### ISSUE THREAD

**Issue by mventimi** [ambiverse-nlu/ambiverse-nlu#21]
We are preparing to process a large corpus (100M documents, 200 GB of data) with AmbiverseNLU, focusing initially on NER/NED using the `entitylinking/analyze` endpoint. Before starting the project, we need to optimize performance in order to (1) process the corpus in a reasonable amount of time and (2) minimize cost. Based on our testing to date, we have some questions about performance:

1. **Benchmarking data**. Is there a standard set of test data you use for benchmarking? If not, we can build something from our data, but are there any specific properties you would recommend such a data set have?

2. **Configuration**. Are there any configuration parameters that we should set to optimize performance? One specific area we thought about is languages--is there any benefit to restricting the application to English if that is the only language we need? If we were to do so, would it be difficult to add additional languages in the future?

3. **Monitoring**. We are monitoring our instances using htop and console logs. Is there any additional performance monitoring capability built into AmbiverseNLU?

4. **Internal parallelization**. The README.md file discusses parallel disambiguation in the context of memory requirements, implying that memory is the limiting factor in running parallel queries. Do CPU and memory usage scale roughly linearly with parallel queries? Is there any advantage to running parallel queries on one large instance, vs. running individual queries on smaller instances?

5. **External parallelization**. We have been testing AmbiverseNLU on AWS EC2 m5n.4xlarge instances, which have 16 cores and 64 GB of RAM, and support 4,750 IOPS. On such a machine, where would you expect AmbiverseNLU to be resource bound? We also did some testing on an m5n.2xlarge instance, which has 16 cores and 32 GB of RAM. It ran about 75% of the transaction rate that the larger instances did, suggesting that we were not totally CPU and/or RAM bound. However, with the lower RAM the instance crashed 1-2 times per hour as described in #19. If we could solve these crashes, this would be a great solution for us, as we could get 75% of the performance for 50% of the price.

6. **Deployment time**. The biggest issue we have with new instances is how long it takes them to spin up. Based on what we have seen so far, we have the following questions:
   - Can we download the database image and specify the local location to PostgreSQL?
   - Can we pre-build the database and store the backup image locally, then restore it on instance start?
   - Are there other things we can do to cache resources locally or pre-build them for either the PostgreSQL or AmbiverseNLU images?

Some of these questions will likely require us to do our own benchmarking--that's fine. I just wanted to make sure we were looking at the right things before doing so. Any guidance is appreciated. In addition, we'd be happy to share our results with you, in order to come up with some typical measurements on AWS instance types.

I know this is a long list of questions. Please let me know if it is helpful to split these up into their own individual issues.

Thanks in advance for the help!

**Comment by hoffart** [comment]
Hi, thanks for the clear structure of questions, sorry for the late answer (busy week). Let me address in order

1. We have been using a random sample of news articles for benchmarking internally, and on average it took 6 seconds per article per processing thread. For the best benchmarking setup, if you already have data you want to process, I would suggest randomly sampling from this data. 
2. There are some configurations (like maximum number of candidates to consider) which will improve the performance at the expense of accuracy. There is no config that simply speeds up without any downside. Are you willing to trade accuracy for performance? And, yes, restricting to English-only will improve performance, because less data is retrieved from the db (all languages are stored and retrieved at once).
3. Monitoring by piping logs into ElasticSearch is a possibility, the service has JSON log output capabilities. Nothing built-in though.
4. Everything scales roughly linearly, as we are parallelizing by documents. However, some models (e.g. for fuzzy lookup or NER) are shared among threads, so running larger instances will give some gain.
5. One of the components of NED is actually quadratic in the number of candidate entities (which is correlated to number of mentions). This might explain the crashes, if you have an unfortunate combination of running multiple large documents in parallel. Our rule of thumb is to run 1 CPU for 4-8 GB of ram, so 8 cores with 64 should be the sweet spot. We have been running this setup with 7-8 threads.
6. Yes, you should set up the database independently of the actual compute nodes. Look at the docker-compose, it is actually spinning up two docker containers. A single Postgres instance should be able to handle multiple compute instances in parallel. For best performance, I recommend running a Cassandra container on the same host. Also, you should be able to link the database containers to a local directory, so that the data stays. We created the docker images for the db with a focus on "quickly testing" AmbiverseNLU, not necessarily for continued deployment.

Hope this helps, please let me know if you have further questions!
